src/libaktualizr/primary/aktualizr_test.cc

```diff
@@ -492,6 +492,355 @@ TEST(Aktualizr, FullWithUpdatesNeedReboot) {
+      "UpdateCheckComplete",    "DownloadProgressReport", "DownloadTargetComplete", "DownloadProgressReport",
+      "DownloadTargetComplete", "AllDownloadsComplete",   "InstallStarted",         "InstallTargetComplete",
+      "InstallStarted",         "InstallTargetComplete",  "AllInstallsComplete",    "PutManifestComplete"};
+
+  std::vector<std::string> expected_report_order = {
+      "EcuDownloadStarted",     "EcuDownloadCompleted",     "EcuDownloadStarted",     "EcuDownloadCompleted",
+      "EcuInstallationStarted", "EcuInstallationCompleted", "EcuInstallationStarted", "EcuInstallationCompleted"};
+
+  const std::string primary_ecu_id = "CA:FE:A6:D2:84:9D";
+  const std::string secondary_ecu_id = "secondary_ecu_serial";
+
+  {
+    fiu_init(0);
+    fiu_enable("fake_package_install", 1, nullptr, 0);
+
+    Aktualizr aktualizr(conf, storage, http_server_mock);
+    EventHandler event_hdlr(aktualizr);
+    aktualizr.Initialize();
+
+    // verify currently installed version
+    std::vector<Uptane::Target> installed_versions;
+    size_t current_version{0};
+    size_t pending_version{0};
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+
+    // for some reason there is no any installed version at initial Aktualizr boot/run
+    // IMHO it should return currently installed version
+    ASSERT_TRUE(installed_versions.empty());
+    ASSERT_EQ(pending_version, 0);
+    ASSERT_EQ(current_version, 0);
+
+    aktualizr.UptaneCycle();
+    aktualizr.uptane_client_->completeInstall();
+
+    ASSERT_TRUE(event_hdlr.checkRecievedEvents(expected_event_order));
+    ASSERT_FALSE(aktualizr.uptane_client_->hasPendingUpdates());
+    ASSERT_TRUE(http_server_mock->checkRecievedReports(expected_report_order));
+
+    data::InstallationResult dev_installation_res;
+    std::string report;
+    std::string correlation_id;
+
+    // `device_installation_result` and `ecu_installation_results` are cleared
+    // at UptaneCycle()->putManifest() once a device manifest is successfully sent to a server
+    ASSERT_FALSE(storage->loadDeviceInstallationResult(&dev_installation_res, &report, &correlation_id));
+
+    std::vector<std::pair<Uptane::EcuSerial, data::InstallationResult>> ecu_installation_res;
+    EXPECT_FALSE(storage->loadEcuInstallationResults(&ecu_installation_res));
+    EXPECT_EQ(ecu_installation_res.size(), 0);
+
+    ASSERT_TRUE(
+        storage->loadInstalledVersions(primary_ecu_id, &installed_versions, &current_version, &pending_version));
+    // it says that no any installed version found,
+    // which is, on one hand is correct since installation of the found update failed hence nothing was installed,
+    // on the other hand some version should have been installed prior to the failed update
+    EXPECT_EQ(installed_versions.size(), 0);
+    EXPECT_EQ(current_version, 0);
+    EXPECT_EQ(pending_version, 0);
```

Comment:
Yes, this is a bit odd, but since nothing got stored to that table, we should expect it to be empty. So for better or worse, for how it works now, this is probably correct.

Reply:
ok